rewrite-csharp/src/main/java/org/openrewrite/csharp/tree/Cs.java

```diff
@@ -1944,4 +1941,522 @@ public String toString() {
+        @Nullable
+        @NonFinal
+        transient WeakReference<Padding> padding;
+
+        @With
+        @Getter
+        @EqualsAndHashCode.Include
+        UUID id;
+
+        @With
+        @Getter
+        Space prefix;
+
+        @With
+        @Getter
+        Markers markers;
+
+        JContainer<AllowsConstraint> expressions;
+
+        public List<AllowsConstraint> getExpressions() {
+            return expressions.getElements();
+        }
+
+        public AllowsConstraintClause withExpressions(List<AllowsConstraint> expressions) {
+            return getPadding().withExpressions(JContainer.withElements(this.expressions, expressions));
+        }
+
+        @Override
+        public <P> J acceptCSharp(CSharpVisitor<P> v, P p) {
+            return v.visitAllowsConstraintClause(this, p);
+        }
+
+        public Padding getPadding() {
+            Padding p;
+            if (this.padding == null) {
+                p = new Padding(this);
+                this.padding = new WeakReference<>(p);
+            } else {
+                p = this.padding.get();
+                if (p == null || p.t != this) {
+                    p = new Padding(this);
+                    this.padding = new WeakReference<>(p);
+                }
+            }
+            return p;
+        }
+
+        @RequiredArgsConstructor
+        public static class Padding {
+            private final AllowsConstraintClause t;
+
+            public JContainer<AllowsConstraint> getExpressions() {
+                return t.expressions;
+            }
+
+            public AllowsConstraintClause withExpressions(JContainer<AllowsConstraint> expressions) {
+                return t.expressions == expressions ? t : new AllowsConstraintClause(t.id, t.prefix, t.markers, expressions);
+            }
+        }
+    }
+
+    /**
+     * Represents a ref struct constraint in a where clause.
+     * Example: where T : allows ref struct
+     */
+    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @RequiredArgsConstructor
+    final class RefStructConstraint implements Cs, AllowsConstraint {
+        @With
+        @Getter
+        @EqualsAndHashCode.Include
+        UUID id;
+
+        @With
+        @Getter
+        Space prefix;
+
+        @With
+        @Getter
+        Markers markers;
+
+        @Override
+        public <P> J acceptCSharp(CSharpVisitor<P> v, P p) {
+            return v.visitRefStructConstraint(this, p);
+        }
+    }
+
+
+    /**
+     * Represents a class/struct constraint in a where clause.
+     * Example: where T : class, where T : struct
+     */
+    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @RequiredArgsConstructor
+    final class ClassOrStructConstraint implements Cs, TypeParameterConstraint {
```

Comment:
Should we maybe declare all these constraints as nested classes rather than top-level?